File: ScapesCore/src/main/java/org/tobi29/scapes/entity/client/MobPlayerClientMain.java
package org.tobi29.scapes.entity.client;

import java8.util.Optional;
import java8.util.function.Consumer;
import org.tobi29.scapes.block.AABBElement;
import org.tobi29.scapes.block.BlockType;
import org.tobi29.scapes.block.TerrainTexture;
import org.tobi29.scapes.chunk.WorldClient;
import org.tobi29.scapes.chunk.terrain.TerrainClient;
import org.tobi29.scapes.client.ScapesClient;
import org.tobi29.scapes.client.connection.ClientConnection;
import org.tobi29.scapes.client.input.InputMode;
import org.tobi29.scapes.client.states.GameStateGameMP;
import org.tobi29.scapes.engine.gui.Gui;
import org.tobi29.scapes.engine.utils.Pool;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.AABB;
import org.tobi29.scapes.engine.utils.math.FastMath;
import org.tobi29.scapes.engine.utils.math.Frustum;
import org.tobi29.scapes.engine.utils.math.PointerPane;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;
import org.tobi29.scapes.engine.utils.math.vector.Vector2f;
import org.tobi29.scapes.engine.utils.math.vector.Vector3;
import org.tobi29.scapes.engine.utils.math.vector.Vector3d;
import org.tobi29.scapes.entity.particle.ParticleEmitterBlock;
import org.tobi29.scapes.packets.PacketInteraction;

import java.util.Iterator;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

public abstract class MobPlayerClientMain extends MobPlayerClient {
    protected final GameStateGameMP game;
    private final Frustum viewField;
    protected Controller controller;
    protected boolean flying;
    protected int swim;
    @SuppressWarnings("CanBeFinal")
    protected double gravitationMultiplier = 1.0, airFriction = 0.2,
            groundFriction = 1.6, wallFriction = 2.0, waterFriction = 8.0,
            stepHeight = 1.0;

    protected MobPlayerClientMain(WorldClient world, Vector3 pos, Vector3 speed,
            AABB aabb, double lives, double maxLives, Frustum viewField,
            String nickname) {
        super(world, pos, speed, aabb, lives, maxLives, nickname);
        this.viewField = viewField;
        game = world.game();
        ScapesClient game = (ScapesClient) this.game.engine().game();
        Consumer<InputMode> inputChange = inputMode -> {
            Gui inputGui = this.game.input();
            inputGui.removeAll();
            inputMode.createInGameGUI(inputGui, world);
            controller = inputMode.playerController(this);
        };
        game.onInputMode(this, inputChange);
        inputChange.accept(game.inputMode());
    }

    protected static Iterator<AABB> collisions(Pool<AABBElement> aabbs) {
        return Streams.of(aabbs).filter(AABBElement::isSolid)
                .map(AABBElement::aabb).iterator();
    }

    protected void updateVelocity(double gravitation, double delta) {
        speed.div(1.0 + airFriction * delta);
        if (inWater) {
            speed.div(1.0 + waterFriction * delta);
        } else {
            if (ground) {
                speed.div(1.0 + groundFriction * delta * gravitation);
            }
            if (slidingWall) {
                speed.div(1.0 + wallFriction * delta);
            }
        }
        speed.plusZ(-gravitation * gravitationMultiplier * delta);
    }

    protected void move(AABB aabb, Pool<AABBElement> aabbs, double goX,
            double goY, double goZ) {
        boolean ground = false;
        boolean slidingWall = false;
        double lastGoZ = aabb.moveOutZ(collisions(aabbs), goZ);
        pos.plusZ(lastGoZ);
        aabb.add(0, 0, lastGoZ);
        if (lastGoZ - goZ > 0) {
            ground = true;
        }
        // Walk
        boolean walking = true;
        while (walking) {
            walking = false;
            if (goX != 0.0) {
                double lastGoX = aabb.moveOutX(collisions(aabbs), goX);
                if (lastGoX != 0.0) {
                    pos.plusX(lastGoX);
                    aabb.add(lastGoX, 0.0, 0.0);
                    goX -= lastGoX;
                    walking = true;
                }
            }
            if (goY != 0.0) {
                double lastGoY = aabb.moveOutY(collisions(aabbs), goY);
                if (lastGoY != 0.0) {
                    pos.plusY(lastGoY);
                    aabb.add(0.0, lastGoY, 0.0);
                    goY -= lastGoY;
                    walking = true;
                }
            }
        }
        // Check collision
        boolean slidingX = goX != 0.0;
        boolean slidingY = goY != 0.0;
        if (slidingX || slidingY) {
            if (stepHeight > 0.0 && (this.ground || inWater)) {
                // Step
                // Calculate step height
                AABB aabbStep = new AABB(aabb).add(goX, 0.0, 0.0);
                double stepX = aabbStep.moveOutZ(collisions(aabbs), stepHeight);
                aabbStep = new AABB(aabb).add(0.0, goY, 0.0);
                double stepY = aabbStep.moveOutZ(collisions(aabbs), stepHeight);
                double step = FastMath.max(stepX, stepY);
                aabbStep = new AABB(aabb).add(goX, goY, step);
                step += aabbStep.moveOutZ(collisions(aabbs), -step);
                // Check step height
                aabbStep.copy(aabb).add(0.0, 0.0, step);
                step = aabb.moveOutZ(collisions(aabbs), step);
                // Attempt walk at new height
                double lastGoX = aabbStep.moveOutX(collisions(aabbs), goX);
                aabbStep.add(lastGoX, 0.0, 0.0);
                double lastGoY = aabbStep.moveOutY(collisions(aabbs), goY);
                // Check if walk was successful
                if (lastGoX != 0.0 || lastGoY != 0.0) {
                    pos.plusX(lastGoX);
                    pos.plusY(lastGoY);
                    aabb.copy(aabbStep).add(0.0, lastGoY, 0.0);
                    pos.plusZ(step);
                } else {
                    // Collide
                    slidingWall = true;
                    if (slidingX) {
                        speed.setX(0.0);
                    }
                    if (slidingY) {
                        speed.setY(0.0);
                    }
                }
            } else {
                // Collide
                slidingWall = true;
                if (slidingX) {
                    speed.setX(0.0);
                }
                if (slidingY) {
                    speed.setY(0.0);
                }
            }
        }
        this.ground = ground;
        this.slidingWall = slidingWall;
    }

    protected void collide(AABB aabb, Pool<AABBElement> aabbs) {
        boolean inWater = false;
        boolean swimming;
        for (AABBElement element : aabbs) {
            if (aabb.overlay(element.aabb)) {
                if (element.collision.isLiquid()) {
                    inWater = true;
                }
            }
        }
        aabb.minZ = FastMath.mix(aabb.minZ, aabb.maxZ, 0.6);
        boolean water = false;
        for (AABBElement element : aabbs) {
            if (aabb.overlay(element.aabb)) {
                if (element.collision.isLiquid()) {
                    water = true;
                }
            }
        }
        if (water) {
            swim++;
            swimming = swim > 1;
        } else {
            swimming = false;
            swim = 0;
        }
        this.inWater = inWater;
        this.swimming = swimming;
    }

    public void updatePosition() {
        positionHandler
                .submitUpdate(entityID, pos.now(), speed.now(), rot.now(),
                        ground, slidingWall, inWater, swimming, true);
    }

    public void onNotice(MobClient notice) {
    }

    @Override
    public void onDamage(double damage) {
        game.client().world().scene().damageShake(damage);
    }

    @Override
    public void move(double delta) {
        updateVelocity(flying ? 0.0 : world.gravity(), delta);
        double goX = FastMath.clamp(speed.doubleX() * delta, -1.0, 1.0);
        double goY = FastMath.clamp(speed.doubleY() * delta, -1.0, 1.0);
        double goZ = FastMath.clamp(speed.doubleZ() * delta, -1.0, 1.0);
        if (flying) {
            pos.plus(new Vector3d(goX, goY, goZ));
        } else {
            AABB aabb = aabb();
            Pool<AABBElement> aabbs = world.terrain().collisions(
                    FastMath.floor(aabb.minX + FastMath.min(goX, 0.0)),
                    FastMath.floor(aabb.minY + FastMath.min(goY, 0.0)),
                    FastMath.floor(aabb.minZ + FastMath.min(goZ, 0.0)),
                    FastMath.floor(aabb.maxX + FastMath.max(goX, 0.0)),
                    FastMath.floor(aabb.maxY + FastMath.max(goY, 0.0)),
                    FastMath.floor(aabb.maxZ + FastMath.max(goZ, stepHeight)));
            move(aabb, aabbs, goX, goY, goZ);
            if (ground) {
                speed.setZ(speed.doubleZ() / (1.0 + 4.0 * delta));
            }
            headInWater = world.terrain().type(pos.intX(), pos.intY(),
                    FastMath.floor(pos.doubleZ() + 0.7)).isLiquid();
            collide(aabb, aabbs);
        }
        positionHandler
                .submitUpdate(entityID, pos.now(), speed.now(), rot.now(),
                        ground, slidingWall, inWater, swimming);
        double lookX = FastMath.cosTable(rot.doubleZ() * FastMath.PI / 180) *
                FastMath.cosTable(rot.doubleX() * FastMath.PI / 180) * 6;
        double lookY = FastMath.sinTable(rot.doubleZ() * FastMath.PI / 180) *
                FastMath.cosTable(rot.doubleX() * FastMath.PI / 180) * 6;
        double lookZ = FastMath.sinTable(rot.doubleX() * FastMath.PI / 180) * 6;
        Vector3 viewOffset = viewOffset();
        viewField.setView(pos.doubleX() + viewOffset.doubleX(),
                pos.doubleY() + viewOffset.doubleY(),
                pos.doubleZ() + viewOffset.doubleZ(), pos.doubleX() + lookX,
                pos.doubleY() + lookY, pos.doubleZ() + lookZ, 0, 0, 1);
        world.entities().filter(entity -> entity instanceof MobClient)
                .forEach(entity -> {
                    MobClient mob = (MobClient) entity;
                    if (viewField.inView(mob.aabb()) > 0) {
                        if (!world.checkBlocked(pos.intX(), pos.intY(),
                                pos.intZ(), entity.pos.intX(),
                                entity.pos.intY(), entity.pos.intZ())) {
                            onNotice(mob);
                        }
                    }
                });
        footStep -= delta;
        if (footStep <= 0.0) {
            footStep = 0.0;
            if (FastMath.max(FastMath.abs((Vector2) speed.now())) > 0.1) {
                int x = pos.intX(), y = pos.intY(), z =
                        FastMath.floor(pos.doubleZ() - 0.1);
                String footSteepSound = world.terrain().type(x, y, z)
                        .footStepSound(world.terrain().data(x, y, z));
                if (footSteepSound.isEmpty() && ground) {
                    z = FastMath.floor(pos.doubleZ() - 1.4);
                    footSteepSound = world.terrain().type(x, y, z)
                            .footStepSound(world.terrain().data(x, y, z));
                }
                if (!footSteepSound.isEmpty()) {
                    Random random = ThreadLocalRandom.current();
                    game.engine().sounds()
                            .playSound(footSteepSound, "sound.World",
                                    0.9f + random.nextFloat() * 0.2f, 1.0f);
                    footStep = 1.0 /
                            FastMath.clamp(FastMath.length(speed.now()), 1.0,
                                    4.0);
                }
            }
        }
        if (invincibleTicks > 0) {
            invincibleTicks--;
        }
    }

    protected void breakParticles(TerrainClient terrain, int amount,
            Vector2 direction) {
        PointerPane pane = block(6, direction);
        if (pane != null) {
            BlockType type = terrain.type(pane.x, pane.y, pane.z);
            Optional<TerrainTexture> tex =
                    type.particleTexture(pane.face, terrain, pane.x, pane.y,
                            pane.z);
            if (tex.isPresent()) {
                TerrainTexture texture = tex.get();
                Vector3 blockPos = new Vector3d(pane.x, pane.y, pane.z);
                ParticleEmitterBlock emitter = world.scene().particles()
                        .emitter(ParticleEmitterBlock.class);
                float r =
                        type.particleColorR(pane.face, terrain, pane.x, pane.y,
                                pane.z);
                float g =
                        type.particleColorG(pane.face, terrain, pane.x, pane.y,
                                pane.z);
                float b =
                        type.particleColorB(pane.face, terrain, pane.x, pane.y,
                                pane.z);
                for (int i = 0; i < amount; i++) {
                    emitter.add(instance -> {
                        Random random = ThreadLocalRandom.current();
                        float size = texture.realSize() * 0.25f;
                        float tx = (random.nextInt(3) + 0.005f) * size;
                        float ty = (random.nextInt(3) + 0.005f) * size;
                        size *= 0.99f;
                        float time = 3.0f;
                        instance.pos.set(blockPos);
                        instance.pos.plus(new Vector3d(random.nextDouble(),
                                random.nextDouble(), random.nextDouble()));
                        instance.speed.set(-1.0f + random.nextFloat() * 2.0f,
                                -1.0f + random.nextFloat() * 2.0f,
                                random.nextFloat() * 2.0f + 1.0f);
                        instance.time = time;
                        instance.dir =
                                random.nextFloat() * (float) FastMath.TWO_PI;
                        instance.textureOffset
                                .set(new Vector2f(texture.realX() + tx,
                                        texture.realY() + ty));
                        instance.textureSize.set(new Vector2f(size, size));
                        instance.setColor(r, g, b, 1.0f);
                    });
                }
            }
        }
    }

    public synchronized void openGui(Gui gui) {
        game.engine().guiStack().add("10-Menu", gui);
        game.setHudVisible(false);
    }

    public Optional<Gui> currentGui() {
        return game.engine().guiStack().get("10-Menu");
    }

    public boolean hasGui() {
        return game.engine().guiStack().has("10-Menu");
    }

    public synchronized boolean closeGui() {
        if (game.engine().guiStack().remove("10-Menu").isPresent()) {
            game.setHudVisible(true);
            world.send(
                    new PacketInteraction(PacketInteraction.CLOSE_INVENTORY));
            return true;
        }
        return false;
    }

    public GameStateGameMP game() {
        return game;
    }

    public ClientConnection connection() {
        return game.client();
    }

    public interface Controller {
        Vector2 walk();

        Vector2 camera(double delta);

        Vector2 hitDirection();

        boolean left();

        boolean right();

        boolean jump();

        int hotbarLeft(int previous);

        int hotbarRight(int previous);
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/entity/client/MobPlayerClientMain.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.entity.client

import org.tobi29.math.AABB3
import org.tobi29.math.Frustum
import org.tobi29.math.threadLocalRandom
import org.tobi29.math.vector.Vector2d
import org.tobi29.math.vector.Vector3d
import org.tobi29.math.vector.add
import org.tobi29.math.vector.length
import org.tobi29.scapes.block.AABBElement
import org.tobi29.scapes.block.BlockType
import org.tobi29.scapes.chunk.WorldClient
import org.tobi29.scapes.chunk.terrain.TerrainClient
import org.tobi29.scapes.client.connection.ClientConnection
import org.tobi29.scapes.client.input.InputModeScapes
import org.tobi29.scapes.client.states.GameStateGameMP
import org.tobi29.scapes.engine.gui.Gui
import org.tobi29.scapes.entity.EntityPhysics
import org.tobi29.scapes.entity.EntityType
import org.tobi29.scapes.entity.ListenerToken
import org.tobi29.scapes.entity.getEntities
import org.tobi29.scapes.entity.particle.ParticleEmitterBlock
import org.tobi29.scapes.packets.PacketInteraction
import org.tobi29.stdex.ThreadLocal
import org.tobi29.stdex.math.TWO_PI
import org.tobi29.stdex.math.clamp
import org.tobi29.stdex.math.floorToInt
import org.tobi29.stdex.math.toRad
import org.tobi29.utils.EventDispatcher
import org.tobi29.utils.ListenerRegistrar
import org.tobi29.utils.Pool
import org.tobi29.utils.forAllObjects
import kotlin.collections.set
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin

abstract class MobPlayerClientMain(
    type: EntityType<*, *>,
    world: WorldClient,
    pos: Vector3d,
    speed: Vector3d,
    aabb: AABB3,
    lives: Double,
    maxLives: Double,
    private val viewField: Frustum,
    nickname: String
) : MobPlayerClient(
    type,
    world, pos, speed, aabb, lives, maxLives, nickname
) {
    val game = world.game
    protected lateinit var input: InputModeScapes
    protected var flying = false
    protected var gravitationMultiplier = 1.0
    protected var airFriction = 0.2
    protected var groundFriction = 1.6
    protected var wallFriction = 2.0
    protected var waterFriction = 8.0
    protected var stepHeight = 1.0
    protected val sendPositionHandler = MobPositionSenderClient(
        registry,
        this.pos.now()
    ) { world.connection.send(it) }
    private var inputEventDispatcherwner: EventDispatcher? = null

    init {
        onDamage[PLAYER_LISTENER_TOKEN] = { damage ->
            world.scene.damageShake(damage)
        }
    }

    fun updatePosition() {
        sendPositionHandler.submitUpdate(
            uuid, pos.now(), speed.now(),
            rot.now(), physicsState.isOnGround, physicsState.slidingWall,
            physicsState.isInWater, physicsState.isSwimming, true
        )
    }

    override fun move(delta: Double) {
        if (flying) {
            val goX = clamp(speed.x * delta, -10.0, 10.0)
            val goY = clamp(speed.y * delta, -10.0, 10.0)
            val goZ = clamp(speed.z * delta, -10.0, 10.0)
            pos.add(Vector3d(goX, goY, goZ))
        } else {
            EntityPhysics.updateVelocity(
                delta, speed, world.gravity,
                gravitationMultiplier, airFriction, groundFriction,
                waterFriction, wallFriction, physicsState
            )
            val aabb = currentAABB()
            val aabbs = AABBS.get()
            EntityPhysics.collisions(
                delta, speed, world.terrain, aabb,
                stepHeight, aabbs
            )
            EntityPhysics.move(
                delta, pos, speed, aabb, stepHeight,
                physicsState, aabbs
            )
            if (isOnGround) {
                speed.setZ(speed.z / (1.0 + 4.0 * delta))
            }
            EntityPhysics.collide(delta, aabb, aabbs, physicsState) {}
            isHeadInWater = world.terrain.type(
                pos.x.floorToInt(),
                pos.y.floorToInt(), (pos.z + 0.7).floorToInt()
            ).isLiquid
            aabbs.reset()
            aabbs.forAllObjects { it.collision = BlockType.STANDARD_COLLISION }
        }
        sendPositionHandler.submitUpdate(
            uuid, pos.now(), speed.now(),
            rot.now(), physicsState.isOnGround, physicsState.slidingWall,
            physicsState.isInWater, physicsState.isSwimming
        )
        val lookX = cos(rot.z.toRad()) *
                cos(rot.x.toRad()) * 6.0
        val lookY = sin(rot.z.toRad()) *
                cos(rot.x.toRad()) * 6.0
        val lookZ = sin(rot.x.toRad()) * 6
        val viewOffset = viewOffset()
        viewField.setView(
            pos.x + viewOffset.x,
            pos.y + viewOffset.y,
            pos.z + viewOffset.z,
            pos.x + lookX, pos.y + lookY,
            pos.z + lookZ, 0.0, 0.0, 1.0
        )
        world.getEntities(
            viewField
        ).filterIsInstance<MobClient>().forEach { entity ->
            val pos2 = entity.getCurrentPos()
            if (!world.checkBlocked(
                    pos.x.floorToInt(),
                    pos.y.floorToInt(),
                    pos.z.floorToInt(),
                    pos2.x.floorToInt(),
                    pos2.y.floorToInt(),
                    pos2.z.floorToInt()
                )) {
                notice(entity)
            }
        }
        footStep -= delta
        if (footStep <= 0.0) {
            footStep = 0.0
            val currentSpeed = speed()
            if (max(
                    abs(currentSpeed.x),
                    abs(currentSpeed.y)
                ) > 0.1) {
                val x = pos.x.floorToInt()
                val y = pos.y.floorToInt()
                val block = world.terrain.block(
                    x, y,
                    (pos.z - 0.1).floorToInt()
                )
                var footStepSound = world.terrain.type(block).footStepSound(
                    world.terrain.data(block)
                )
                if (footStepSound == null && isOnGround) {
                    val blockBottom = world.terrain.block(
                        x, y,
                        (pos.z - 1.4).floorToInt()
                    )
                    footStepSound = world.terrain.type(
                        blockBottom
                    ).footStepSound(
                        world.terrain.data(blockBottom)
                    )
                }
                if (footStepSound != null) {
                    val random = threadLocalRandom()
                    game.engine.sounds.playSound(
                        footStepSound, "sound.World",
                        0.9 + random.nextDouble() * 0.2, 1.0
                    )
                    footStep = 1.0 / clamp(speed.now().length(), 1.0, 4.0)
                }
            }
        }
        if (invincibleTicks > 0) {
            invincibleTicks--
        }
    }

    protected fun breakParticles(
        terrain: TerrainClient,
        amount: Int,
        direction: Vector2d
    ) {
        val pane = block(6.0, direction)
        if (pane != null) {
            val block = terrain.block(pane.x, pane.y, pane.z)
            val type = terrain.type(block)
            val data = terrain.data(block)
            val texture = type.particleTexture(
                pane.face, terrain, pane.x,
                pane.y, pane.z, data
            )
            if (texture != null) {
                val blockPos = Vector3d(
                    pane.x.toDouble(), pane.y.toDouble(),
                    pane.z.toDouble()
                )
                val emitter = world.scene.particles().emitter(
                    ParticleEmitterBlock::class.java
                )
                val friction = type.particleFriction(
                    pane.face, terrain,
                    pane.x, pane.y, pane.z, data
                )
                val r = type.particleColorR(
                    pane.face, terrain, pane.x, pane.y,
                    pane.z, data
                )
                val g = type.particleColorG(
                    pane.face, terrain, pane.x, pane.y,
                    pane.z, data
                )
                val b = type.particleColorB(
                    pane.face, terrain, pane.x, pane.y,
                    pane.z, data
                )
                for (i in 0 until amount) {
                    emitter.add { instance ->
                        val random = threadLocalRandom()
                        val time = 3.0f
                        instance.pos.set(blockPos)
                        instance.pos.add(
                            Vector3d(
                                random.nextDouble(),
                                random.nextDouble(), random.nextDouble()
                            )
                        )
                        instance.speed.setXYZ(
                            -1.0 + random.nextDouble() * 2.0,
                            -1.0 + random.nextDouble() * 2.0,
                            random.nextDouble() * 2.0 + 1.0
                        )
                        instance.time = time
                        instance.friction = friction
                        instance.dir = random.nextFloat() * TWO_PI.toFloat()
                        instance.setTexture(texture, random)
                        instance.setColor(r, g, b, 1.0f)
                    }
                }
            }
        }
    }

    @Synchronized
    fun openGui(gui: Gui) {
        game.engine.guiStack.add("10-Menu", gui)
        game.setHudVisible(false)
    }

    fun currentGui(): Gui? {
        return game.engine.guiStack["10-Menu"]
    }

    fun hasGui(): Boolean {
        return game.engine.guiStack.has("10-Menu")
    }

    @Synchronized
    fun closeGui(): Boolean {
        if (game.engine.guiStack.remove("10-Menu") != null) {
            game.setHudVisible(true)
            world.send(
                PacketInteraction(
                    world.plugins.registry,
                    PacketInteraction.CLOSE_INVENTORY, 0
                )
            )
            return true
        }
        return false
    }

    fun game(): GameStateGameMP {
        return game
    }

    fun connection(): ClientConnection {
        return game.client()
    }

    @Synchronized
    fun setInputMode(input: InputModeScapes?) {
        inputEventDispatcherwner?.disable()
        game.inputGui.removeAll()
        if (input == null) {
            inputEventDispatcherwner = null
            return
        }
        input.createInGameGUI(game.inputGui, world)
        val inputEventDispatcher = EventDispatcher(game.engine.events) {
            inputMode(input)
        }
        this.input = input
        this.inputEventDispatcherwner = inputEventDispatcher
        inputEventDispatcher.enable()
    }

    protected abstract fun ListenerRegistrar.inputMode(input: InputModeScapes)

    class InputDirectionEvent(val direction: Vector2d)

    class HotbarChangeLeftEvent(val delta: Int) {
        var success: Boolean = true
    }

    class HotbarChangeRightEvent(val delta: Int) {
        var success: Boolean = true
    }

    class HotbarSetLeftEvent(val value: Int) {
        var success: Boolean = true
    }

    class HotbarSetRightEvent(val value: Int) {
        var success: Boolean = true
    }

    class MenuOpenEvent {
        var success: Boolean = true
    }

    class MenuInventoryEvent {
        var success: Boolean = true
    }

    class MenuChatEvent {
        var success: Boolean = true
    }

    companion object {
        private val AABBS = ThreadLocal { Pool { AABBElement() } }
    }
}

private val PLAYER_LISTENER_TOKEN = ListenerToken("Scapes:Player")
